### PR TITLE
Add error handling to rust trade logic

### PR DESCRIPTION
**Pull request by slundqui**
# Resolved Issues
https://github.com/delvtech/hyperdrive/issues/743 (partially)

# Description
Adds the following error checks in rust trades:
- Minimum transaction checks for all long and short trades.
- Adds negative interest checks for open and close shorts.

# Review Checklists

Please check each item **before approving** the pull request. While going
through the checklist, it is recommended to leave comments on items that are
referenced in the checklist to make sure that they are reviewed. If there are
multiple reviewers, copy the checklists into sections titled `## [Reviewer Name]`.
If the PR doesn't touch Solidity and/or Rust, the corresponding checklist can
be removed.

### Rust

- [ ] **Testing**
    - [x] Are there new or updated unit or integration tests?
    - [ ] Do the tests cover the happy paths?
    - [x] Do the tests cover the unhappy paths?
    - [ ] Are there an adequate number of fuzz tests to ensure that we are
          covering the full input space?
    - [ ] If matching Solidity behavior, are there differential fuzz tests that
          ensure that Rust matches Solidity?
          
### Rust

- [x] **Testing**
    - [x] Are there new or updated unit or integration tests?
    - [x] Do the tests cover the happy paths?
    - [x] Do the tests cover the unhappy paths?
    - [x] Are there an adequate number of fuzz tests to ensure that we are
          covering the full input space?
    - [x] If matching Solidity behavior, are there differential fuzz tests that
          ensure that Rust matches Solidity?

### CODE REVIEW

**Comment by slundqui** [comment]
This PR still needs tests to make sure these checks are being hit in these cases. Would love some help coming up with known trades that result in insufficient liquidity

File: crates/hyperdrive-math/src/short/open.rs [hunk 1]
```diff
@@ -34,19 +34,35 @@ impl State {
...
         // If the open share price hasn't been set, we use the current share
         // price, since this is what will be set as the checkpoint share price
         // in the next transaction.
         if open_vault_share_price == fixed!(0) {
             open_vault_share_price = self.vault_share_price();
         }
 
+        // TODO solidity uses mulUp here, should we do that here as well?
+        let share_reserves_delta_in_shares =
+            self.vault_share_price() * self.short_principal(short_amount)?;
```
Comment (ryangoree) [inline crates/hyperdrive-math/src/short/open.rs:51]: Looks like it was changed to round up a couple months ago. We should probably change this and the calc below to round up to remain at parity and overestimate the proceeds for the negative interest check.

https://github.com/delvtech/hyperdrive/pull/736/files#diff-114831fb0435b556f3b41af29e9f3b9e808cbd0dbd8d3ca504492a09ed4fd6fcR425

Also, I think the name of this variable is a typo, because the number is denominated in base.

File: crates/hyperdrive-math/src/long/close.rs [hunk 2]
```diff
@@ -36,6 +36,13 @@ impl State {
         normalized_time_remaining: F,
     ) -> FixedPoint {
         let bond_amount = bond_amount.into();
+
+        if bond_amount < self.config.minimum_transaction_amount.into() {
+            // TODO would be nice to return a `Result` here instead of a panic.
```
Comment (ryangoree) [inline crates/hyperdrive-math/src/long/close.rs:41]: Agreed. If we don't implement them in this PR, can we add an issue for these so it can be prioritized in the backlog?

Comment (dpaiton) [reply crates/hyperdrive-math/src/long/close.rs:41]: Agreed; leave it as a TODO & make an issue. I'll handle it, I already fixed a few in targeted_long PR and I think I'm going to just do all of them.

**Comment by dpaiton** [comment]
Since you're testing these "unhappy paths" & verifying error handling, I think it makes sense to include a set of tests that assert that there is an error if one tries to trade more than the maximum amount.

**Comment by dpaiton** [comment]
This doesn't fully resolve #743, right? We now have coverage for open/close long/short but not for add/remove liquidity.

**Comment by dpaiton** [comment]
Part of the checklist asks if we have sufficient testing against the solidity contracts. One way we could do this is to fuzz a large number of open & close trade amounts, with fuzz bounds that are outside the range of allowable trades (too big & too small). Then the test confirms that if rust allows it, solidity allows it and vice versa. 

There's a python example of this in #743 . I think it would be good to add these tests in this PR.

**Comment by slundqui** [comment]
> This doesn't fully resolve #743, right? We now have coverage for open/close long/short but not for add/remove liquidity.

We currently don't have an add/remove liquidity in rust:
https://github.com/delvtech/hyperdrive/issues/873
https://github.com/delvtech/hyperdrive/issues/874
https://github.com/delvtech/hyperdrive/issues/872

**Comment by slundqui** [comment]
> > > This doesn't fully resolve #743, right? We now have coverage for open/close long/short but not for add/remove liquidity.
> > 
> > 
> > We currently don't have an add/remove liquidity in rust: #873 #874 #872
> 
> Got it; I updated #743. What about the max fns? Should we move that to a new issue or will you address it here?

Up to you if you want to keep these in that issue, but I was going to group error handling while tackling this issue:
https://github.com/delvtech/hyperdrive/issues/871

**Comment by dpaiton** [comment]
> > > > This doesn't fully resolve #743, right? We now have coverage for open/close long/short but not for add/remove liquidity.
> > > 
> > > 
> > > We currently don't have an add/remove liquidity in rust: #873 #874 #872
> > 
> > 
> > Got it; I updated #743. What about the max fns? Should we move that to a new issue or will you address it here?
> 
> Up to you if you want to keep these in that issue, but I was going to group error handling while tackling this issue: #871

sounds good to me, I'll move the task over.

File: crates/hyperdrive-math/src/short/close.rs [hunk 3]
```diff
@@ -75,12 +91,30 @@ impl State {
...
         let share_reserves_delta = self
             .calculate_close_short_flat_plus_curve(bond_amount, normalized_time_remaining)
             + self.close_short_curve_fee(bond_amount, normalized_time_remaining)
             + self.close_short_flat_fee(bond_amount, normalized_time_remaining);
 
+        // Throw an error if closing the short would result in negative interest.
+        let ending_spot_price = {
+            let mut state: State = self.clone();
+            state.info.bond_reserves -= bond_amount.into();
+            state.info.share_reserves += share_reserves_delta.into();
```
[the remark this answers is not in the document]
Comment (slundqui) [reply crates/hyperdrive-math/src/short/close.rs]: Reworked to include the two negative interest checks, one of which requires accounting for governance fee

**Comment by slundqui** [comment]
> Part of the review checklist asks if we have sufficient testing against the solidity contracts. One way we could do this is to fuzz a large number of open & close trade amounts, with fuzz bounds that are outside the range of allowable trades (too big & too small). Then the test confirms that if rust allows it, solidity allows it and vice versa.
> 
> There's a python example of this in #743 . I think it would be good to add these tests in this PR.

Digging down into this, this is a pretty big undertaking, as our current integration tests (1) doesn't exist for open calls, and (2) only tests against the underlying hyperdrive math functions (instead of the long/short interface), which bypasses both solidity's and rust's fee and error check calculations. Is this something we can split up into a separate PR, and is this critical for our deadlines?

**Comment by dpaiton** [comment]
> > Part of the review checklist asks if we have sufficient testing against the solidity contracts. One way we could do this is to fuzz a large number of open & close trade amounts, with fuzz bounds that are outside the range of allowable trades (too big & too small). Then the test confirms that if rust allows it, solidity allows it and vice versa.
> > There's a python example of this in #743 . I think it would be good to add these tests in this PR.
> 
> Digging down into this, this is a pretty big undertaking, as our current integration tests (1) doesn't exist for open calls, and (2) only tests against the underlying hyperdrive math functions (instead of the long/short interface), which bypasses both solidity's and rust's fee and error check calculations. Is this something we can split up into a separate PR, and is this critical for our deadlines?

Agreed, sounds like a new PR to me. Can you make the issue?

**Comment by slundqui** [comment]
> > > Part of the review checklist asks if we have sufficient testing against the solidity contracts. One way we could do this is to fuzz a large number of open & close trade amounts, with fuzz bounds that are outside the range of allowable trades (too big & too small). Then the test confirms that if rust allows it, solidity allows it and vice versa.
> > > There's a python example of this in #743 . I think it would be good to add these tests in this PR.
> > 
> > 
> > Digging down into this, this is a pretty big undertaking, as our current integration tests (1) doesn't exist for open calls, and (2) only tests against the underlying hyperdrive math functions (instead of the long/short interface), which bypasses both solidity's and rust's fee and error check calculations. Is this something we can split up into a separate PR, and is this critical for our deadlines?
> 
> Agreed, sounds like a new PR to me. Can you make the issue?

Issue made here:
https://github.com/delvtech/hyperdrive/issues/937

File: crates/hyperdrive-math/src/short/close.rs [hunk 4]
```diff
@@ -78,16 +113,58 @@ impl State {
...
+            state.info.share_reserves += share_curve_delta_with_fees.into();
+            state.calculate_spot_price()
+        };
+        if share_curve_delta_with_fees_spot_price > fixed!(1e18) {
+            // TODO would be nice to return a `Result` here instead of a panic.
+            panic!("InsufficientLiquidity: Negative Interest");
+        }
+
+        // Now calculate short proceeds
+        // TODO we've already calculated a couple of internal variables needed by this function,
```
Comment (dpaiton) [inline crates/hyperdrive-math/src/short/close.rs:155]: agreed; can you make a high priority issue for this so we keep track of it?

File: crates/hyperdrive-math/src/short/fees.rs [hunk 5]
```diff
@@ -40,6 +40,19 @@ impl State {
             * bond_amount.mul_div_down(normalized_time_remaining, self.vault_share_price())
     }
 
+    // Calculate the curve portion of the governance fee for close shorts
+    // NOTE: Round down to underestimate the governance curve fee
+    // TODO: avoid duplicate calculation of close short curve fee
```
Comment (dpaiton) [inline crates/hyperdrive-math/src/short/fees.rs:45]: you can include this in the other issue on making `calculate_close_short_flat_plus_curve` more efficient

**Comment by dpaiton** [comment]
I think #743 is not resolved yet. Correct me if I'm wrong, but I think we are still missing:

- [ ] upper-bound check & test on close long
  - what if I try to close for more than I have? Is this just a balanceOf check? What about closing more than the pool has? 
- [ ] upper-bound check & test on close short
- [ ] open short negative interest test (check is there; but the unhappy path test is not)

Apologies if you've already said why you didn't do these 😬 

It can be a pain to handle all of the tasks in a monolith issue like that, so I'm fine with breaking out smaller tasks, closing #743, and merging this PR before continuing with the remaining ones.

**Review by dpaiton — APPROVED** [review]
